src/algorithms/polygon_visibility/naive.rs

```diff
@@ -0,0 +1,198 @@
+//
+// Algorithm overview:
+//   1. For each vertex in the polygon (including outer boundary and any holes):
+//     1.A Shoot a right-leaning SoS ray towards the vertex and record the intersection
+//         point closest to the origin. Note: The SoS ray might not hit the vertex.
+//     1.B Shoot a left-leaning SoS ray towards the vertex and record the intersection
+//         point closest to the origin. This point may or may not be the same as the point
+//         in step 1.A.
+//   2. Sort all the intersection points counter-clockwise around the point of origin.
+//   3. If two points are at the same angle then sort according to the leaning of the
+//      ray that generated the intersection point. Ie. a point from a left-leaning ray
+//      should be Ordering::Greater than a point from a right-leaning ray.
+//
+// Pseudo code:
+//   for vertex in poly.vertex {
+//     let left_intersections = ... use left-leaning SoS ray ...;
+//     let right_intersections = ... use right-leaning SoS ray ...;
+//     let left_closest = ...;
+//     let right_closest = ...;
+//     output.push( (SoS::CounterClockWise, left_closest));
+//     output.push( (SoS::ClockWise, right_closest));
+//   }
+//   output.sort_by(|a,b| origin.ccw_cmp_around(a,b).then(cmp by sos) );
+//   Polygon::new(output)
+//
+// Test cases:
+//
+//   Input            Output
+//
+//  /-----\ /----\  /-----\ /----\
+//  |     | |    |  |     | |    |
+//  |  x  \-/ /--/  |  x  \---\--/
+//  |         |     |         |
+//  \---------/     \---------/
+//
+//
+//  /-----\ /----\  /-----\ /----\
+//  |     | |    |  |     | |    |
+//  |  x  \-/    |  |  x  \------\
+//  |            |  |            |
+//  \------------/  \------------/
+//
+//  /---------\     /-\     /-\
+//  |   /-\   |     |  \   /  |
+//  |   \-/   |     |   \ /   |
+//  |    x    |     |    x    |
+//  \---------/     \---------/
+
+pub fn get_visibility_polygon<T>(point: &Point<T, 2>, polygon: &Polygon<T>) -> Option<Polygon<T>>
+where
+  T: PolygonScalar,
+{
+  //check whether the point is in the polygon or not
+  if polygon.locate(point) == PointLocation::Outside {
+    return Option::None;
+  }
+
+  let polygon_segments: Vec<LineSegment<T, 2>> = polygon
+    .points
+    .windows(2)
+    .map(|pair| {
+      LineSegment::new(
+        EndPoint::Inclusive(pair[0].clone()),
+        EndPoint::Inclusive(pair[1].clone()),
+      )
+    })
+    .collect();
+
+  let mut polygon_points = Vec::new();
+  for vertex in &polygon.points {
+    let right_sos = HalfLineSoS {
+      line: LineSoS::new_through(point.clone(), vertex.clone()),
+      lean: SoS::ClockWise,
+    };
+    let left_sos = HalfLineSoS {
+      line: LineSoS::new_through(point.clone(), vertex.clone()),
+      lean: SoS::CounterClockWise,
+    };
+
+    let mut right_intersections = Vec::new();
+    let mut left_intersections = Vec::new();
+
+    polygon_segments.iter().for_each(|segment| {
+      if right_sos.intersect(segment.as_ref()).is_some() {
```

Comment:
If we use `DirectedEdge` from `polygon.iter_boundary_edges()` then we don't have to use `as_ref()`. The code can look like this:
```rust
if right_sos.intersect(edge).is_some() {
  right_intersections.push(get_intersection(&right_sos, edge.into()));
}
```
This requires that we change `get_intersection` to take a `LineSegmentView`.

Reply:
I found that I have to call clone twice to copy a point from the LineSegmentView. I don't understand where the double reference came from but had a bad feeling about calling clone twice so I used &DirectedEdge. Am I missing something?

Reply:
Hm, this is a problem with the clunky edge structures. They are not as easy-to-use as I want them to be. Using `&DirectedEdge` is fine. I'll work on better, more consistent edge structures later.